src/smartHandler.ts

```diff
@@ -190,41 +256,69 @@ export class SMARTHandler implements Authorization {
+                resourceType,
+            );
+            if (validOperations.includes(operation)) return true;
+        }
+        return false;
+    }
+
+    private getValidOperationsForScope(
+        scope: string,
+        reqOperation: TypeOperation | SystemOperation,
+        reqResourceType?: string,
+    ): (TypeOperation | SystemOperation)[] {
+        const { scopeRule } = this.config;
+        let match = scope.match(SMARTHandler.LAUNCH_SCOPE_REGEX);
+        if (!match) {
+            match = scope.match(SMARTHandler.CLINICAL_SCOPE_REGEX);
+        }
+        let validOperations: (TypeOperation | SystemOperation)[] = [];
+        if (match !== null) {
+            const { scopeType } = match.groups!;
+            if (scopeType === 'launch') {
+                const { launchType } = match.groups!;
+                // TODO: should launch have access to only certain resourceTypes?
+                if (['patient', 'encounter'].includes(launchType)) {
+                    validOperations = scopeRule[scopeType][<LaunchType>launchType];
+                } else if (!launchType) {
+                    validOperations = scopeRule[scopeType].launch;
+                }
+            } else if (['patient', 'user', 'system'].includes(scopeType)) {
+                const { scopeResourceType, accessType } = match.groups!;
+                if (reqResourceType) {
+                    if (scopeResourceType === '*' || scopeResourceType === reqResourceType) {
+                        validOperations = this.getValidOperationsForScopeGivenScopeRule(
+                            <ScopeType>scopeType,
+                            accessType,
+                        );
                     }
-                } else if (['patient', 'user', 'system'].includes(scopeType)) {
-                    const { scopeResourceType, accessType } = match.groups!;
-                    if (resourceType) {
-                        if (scopeResourceType === '*' || scopeResourceType === resourceType) {
-                            validOperations = this.getValidOperationsForScope(<ScopeType>scopeType, accessType);
-                        }
-                    } else if (scopeResourceType === '*') {
-                        validOperations = this.getValidOperationsForScope(<ScopeType>scopeType, accessType);
+                }
+                // 'search-system' and 'history-system' request operation requires '*' for scopeResourceType
+                else if (['search-system', 'history-system'].includes(reqOperation)) {
+                    if (scopeResourceType === '*') {
+                        validOperations = this.getValidOperationsForScopeGivenScopeRule(
+                            <ScopeType>scopeType,
+                            accessType,
+                        );
                     }
+                } else {
+                    validOperations = this.getValidOperationsForScopeGivenScopeRule(<ScopeType>scopeType, accessType);
                 }

```

Comment:
I still struggle to see why we want add this

if
1. 'patient', 'user', 'system' scope 
1. the request resource is null 
1. it is not a system history/search

If there is no resource defined shouldn't we require the scopeResourceType =='*'? If this is for batch/transaction could we add an edge-case that way instead of search?

Reply:
That's a good question. It's structured this way to allow for two different use case. The first case is the `batch/transaction` use case.

The second case is when `getValidOperationsForScope` is called by [getAllowedResourceTypesForOperation](https://github.com/awslabs/fhir-works-on-aws-authz-smart/blob/ddd77d0163f2b3b14eab41e83f92607f3495bc2c/src/smartHandler.ts#L165). In this second use case we have the scopes and the operation as provided by `AllowedResourceTypesForOperationRequest`. However, we do not have a `resourceType`. Therefore it could be possible where `scope = user/Patient.read` and `operation = search-type`. In this scenario we would still want to get the `validOperations` for that `scope` and `operation`.

Reply:
Interesting Should we change `AllowedResourceTypesForOperationRequest` to include the resourceType?

Reply:
Looks like it already is in requesT: https://github.com/awslabs/fhir-works-on-aws-authz-smart/pull/7/files#diff-7f9bca42e8ef8f0cc4b82957be62aa94575c0886947d59b067cd8f1eaea5ba6eR166

Reply:
For a concrete use-case that I am not sure we are covering say we have rule as:
`{patient:{read:['transaction', 'read']}}`
and a bundle request:
`POST ehr.com/`
with scopes:
`patient/Patient.read`

For this settings we would want to allow the first pass from `verifyaccesstoken` method (checking if the scope gives transaction permission) then in isBundleAuthorized we want to go 1 by 1 to verify if the contents are permitted

Reply:
I guess is there any concern just to make this:
```suggestion
                } else if (['bundle',...].includes(reqOperation)) {
                    validOperations = this.getValidOperationsForScopeGivenScopeRule(<ScopeType>scopeType, accessType);
                }

```